editor/test/integration/test_util.clj

```diff
@@ -675,6 +676,56 @@
 (defn outline [root path]
   (get-in (g/node-value root :node-outline) (interleave (repeat :children) path)))
 
+(defrecord OutlineItemIterator [root path]
+  outline/ItemIterator
+  (value [_this]
+    (outline root path))
+  (parent [_this]
+    (when (not (empty? path))
+      (->OutlineItemIterator root (butlast path)))))
+
+(defn- make-outline-item-iterator [root-node-id outline-path]
+  {:pre [(g/node-id? root-node-id)
+         (vector? outline-path)
+         (every? nat-int? outline-path)]}
+  (->OutlineItemIterator root-node-id outline-path))
+
+(defn outline-copy
+  "Return a serialized a data representation of the specified parts of the
+  edited scene. The returned value could be placed on the clipboard and later
+  pasted into an edited scene."
+  [project root-node-id & outline-paths]
+  (let [item-iterators (mapv #(make-outline-item-iterator root-node-id %) outline-paths)]
+    (outline/copy project item-iterators)))
+
+(defn outline-cut!
+  "Cut the specified parts of the edited scene and return a data representation
+  of the cut nodes that can be placed on the clipboard."
+  [project root-node-id outline-path & mode-outline-paths]
+  (let [item-iterators (into [(make-outline-item-iterator root-node-id outline-path)]
+                             (map #(make-outline-item-iterator root-node-id %))
+                             mode-outline-paths)]
+    (outline/cut! project item-iterators)))
+
+(defn outline-paste!
+  "Paste the copied-data into the edited scene, similar to how the user would
+  perform this operation in the editor."
+  ([project root-node-id outline-path copied-data]
+   (outline-paste! project root-node-id outline-path copied-data nil))
+  ([project root-node-id outline-path copied-data select-fn]
+   (let [item-iterator (make-outline-item-iterator root-node-id outline-path)]
+     (assert (outline/paste? project item-iterator copied-data))
+     (outline/paste! project item-iterator copied-data select-fn))))
+
+(defn outline-duplicate!
+  "Simulate a copy-paste operation of the specified node in-place, typically how
+  a user might duplicate an element in the edited scene."
+  ([project root-node-id outline-path]
+   (outline-duplicate! project root-node-id outline-path nil))
+  ([project root-node-id outline-path select-fn]
+   (let [copied-data (outline-copy project root-node-id outline-path)]
+     (outline-paste! project root-node-id outline-path copied-data select-fn))))
+
```

Comment:
Curiously we did not have helper functions for simulating copy-paste operations in `test-util`. So named Gui resources are the only clipboard operations covered by tests at the moment. 😬

There are some general tests of this for the outline view, but nothing really exercises the various `:tx-attach-fns` of various node types. 😞